tincr/io/library/genCellLibrary.tcl

```diff
@@ -300,275 +727,368 @@ proc process_macrocell {c s fo} {
+# @return The type of cell pin
+proc get_pin_type { pin } {
+
+    if { [get_property IS_CLEAR $pin] } {
+        return "CLEAR"
+    } elseif { [get_property IS_CLOCK $pin] } {
+        return "CLOCK"
+    } elseif { [get_property IS_ENABLE $pin] } {
+        return "ENABLE"
+    } elseif { [get_property IS_PRESET $pin] } {
+        return "PRESET"
+    } elseif { [get_property IS_RESET $pin] } {
+        return "RESET"
+    } elseif { [get_property IS_SET $pin] } {
+        return "SET"
+    } elseif { [get_property IS_SETRESET $pin] } {
+        return "SETRESET"
+    } elseif { [get_property IS_WRITE_ENABLE $pin] } {
+        return "WRITE_ENABLE"
+    } else {
+        return "DATA"
     }
 }
 
-#top level function used to create a cell library file used in RapidSmith2
-proc ::tincr::create_xml_cell_library { {part xc7a100t-csg324-3} {filename ""} } {
+## Generates the cell library XML for the specified cell when it is placed
+#   on each site in the site list (each bel within the site is tested). Currently
+#   ignores macro cells.
+#   TODO: revisit if/when macro cells are supported in RapidSmith.
+#
+# @param cell_instance Library cell instance
+# @param site_type_list List of site types to place cell_instance onto
+# @param site_map Dictionary mapping site types to a physical site location
+# @param alternate_only_sites Set of site types that are only alternate sites
+# @param xml_out File handle to the output XML file
+proc write_bel_placement_xml { cell_instance site_type_list site_map alternate_only_sites xml_out } {
+    
+    set lib_cell_name [get_property REF_NAME $cell_instance]
+    puts $xml_out "      <bels>"
 
-    set part_list [split $part "-"]
+    set net_list [list]
+        
+    #print the placement information
+    foreach site_type $site_type_list {
+        set site [dict get $site_map $site_type]
+
+        set is_alternate 0
+        if { [::struct::set contains $alternate_only_sites $site_type] } {
+            set is_alternate 1
+            set_property MANUAL_ROUTING $site_type $site
+        }
+        
+        if {[get_property PRIMITIVE_LEVEL $cell_instance] == "MACRO"} {
+            # process_macrocell $cell_instance $s $xml_out
+        } else {
+            process_leaf_cell $cell_instance $site $xml_out
+        }
+        
+        if { $is_alternate } {
+            reset_property MANUAL_ROUTING $site
+        }
+    }
+    
+    puts $xml_out "      </bels>"
+}
 
+
+## Creates a cell library XML file that can be used by RapidSmith version 2.0.
+#
+# @param filename Optional parameter to specify the generated cell library name. The default name is
+#           "cellLibrary_part.xml"
+# @param part Xilinx FPGA part to generate a cell library for
+# @param threshold Threshold of configurable pin mappings to compute before quitting. (currently not used
+#           but will be used in the future)
+proc ::tincr::create_xml_cell_library { {part xc7a100t-csg324-3} {filename ""} {threshold 100}} {
```

Comment:
How about putting all these XML generation methods to a file `rapidsmith.tcl` to the `tincr.io.design` package.